libraries/chain/block_log.cpp

```diff
@@ -615,6 +621,63 @@ namespace eosio { namespace chain {
             FC_LOG_AND_RETHROW()
          }
 
+         std::vector<char> read_serialized_block_by_num(uint32_t block_num) final {
+            try {
+               uint64_t pos = get_block_pos(block_num);
+
+               if (pos == block_log::npos || !head) {
+                  // Rare case. No need a special code for it.
+                  // Just fall back to the regular `retry_read_block_by_num` and then serialize.
+                  auto ret = retry_read_block_by_num(block_num);
+                  if (ret) {
+                     return fc::raw::pack(*ret);
+                  } else {
+                     return {};
+                  }
+               }
+
+               uint32_t head_block_num = block_header::num_from_id(head->id);
+               EOS_ASSERT(block_num <= head_block_num, block_log_exception,
```

Comment:
Doesn't this assert change the previous behavior of  `connection::blk_send`? 

Before, it would: 
`peer_ilog( this, "fetch block by id returned null, id ${id}", ("id", blkid) );`

Now it will:
`peer_elog( this, "caught assert on fetch_serialized_block_by_id, ${ex}, id ${id}", ("ex", ex.to_string())("id", blkid) );`

I don't think Areg said we had to always use `EOS_ASSERT`, but we had to avoid undefined behavior. In this case having an info message stating that the requested block was not found seems fine.